### PR TITLE
readLines with fread

### PR DESCRIPTION
**Pull request by dselivanov**
It is tricky to use `fread` as replacement for `readLines` for reading non-tabular raw text data. In fact I realised that I actually can do that only after creating this PR.
Consider following example where I want to read text without splitting it into `data.frame`/`data.table`

``` r
txt <- fread('a,b\n ab,cd,ce\n abcdef\n hjkli \n')
# Error in fread("a,b\n ab,cd,ce\n abcdef\n hjkli \n") : 
#  Expected sep (',') but new line, EOF (or other non printing character) ends field 0 when detecting types (   first):  abcdef
#In addition: Warning message:
#In fread("a,b\n ab,cd,ce\n abcdef\n hjkli \n") :
#  Starting data input on line 2 and discarding line 1 because it has too few or too many items to be column names or data: a,b
```

Fortunately (and I realised that only after browsing `fread.c` source code) we can pass `sep = "\n"`.

``` r
txt <- fread('a,b\n ab,cd,ce\n abcdef\n hjkli \n', sep = '\n')
#        a,b
#1: ab,cd,ce
#2:   abcdef
#3:    hjkli
```

Note, that we received  `"hjkli"`, not `" hjkli "`, because `strip.white = TRUE` by default. This also can be misleading.

And the last - most painful. It is not possible to do trick above on windows, where `EOL = '\r\n'`:

``` r
fread("a,b\n ab,cd,ce\n abcdef\n hjkli \n", sep = "\r\n")
#Error in fread("a,b\n ab,cd,ce\n abcdef\n hjkli \n", sep = "\r\n") : 
#  'sep' must be 'auto' or a single character
```

### CODE REVIEW

**Comment by arunsrinivasan** [comment]
Hm, `Hash Sum mismatch` error in travis-ci.. strange.

Looks nice, could you please remove irrelevant (spacing) lines from the PR so that the actual changes are easily visible to have a look at? Thanks.

**Comment by dselivanov** [comment]
see updated PR. Do you mind if I will open separate PR for syntax/readability refinements?

**Comment by arunsrinivasan** [comment]
@dselivanov I've not yet had time to look at the PR, sorry.

On syntax/readability, I'd suggest starting a new thread/issue so as to develop/write data.table package coding style.

**Comment by MichaelChirico** [comment]
this is cool. i'd been using `sep = "^"` for this but always anxious that there will be a `^` hidden in my file one time. good to have a robust alternative.

**Comment by dselivanov** [comment]
@arunsrinivasan see #1532.

**Comment by arunsrinivasan** [comment]
@dselivanov just had a look. Nice! Could you explain in words about `sep = NULL` in `fread.Rd`?

**Comment by MichaelChirico** [comment]
Could also be useful to have some benchmarks to compare performance vis-a-vis `readLines`... just not sure what the proper medium for putting those benchmarks is. On the wiki somewhere? FAQ?

**Comment by dselivanov** [comment]
@arunsrinivasan sure, will update PR soon.

**Comment by jangorecki** [comment]
@dselivanov fread has `data.table` argument which could be used in that context, when `data.table=FALSE` then a vector would be returned.

**Comment by arunsrinivasan** [comment]
@dselivanov same behaviour as always. Default is `data.table`. `data.table=FALSE` returns `data.frame` (1-col).

On second thought, `sep=NULL` is quite confusing. Would it be possible to have the same functionality with `sep="\n"` instead? The logic to detect if eol is `\r\n` or `\n` is already there.. so `sep="\n"` could just use it?

Also, `sep="\n"` could also be added as the last separator in the auto detection phase.

**Comment by dselivanov** [comment]
@arunsrinivasan sure, we can check for `"\n"`. For me `sep = NULL` looks very natural:

1. `sep = NULL` - do not split/parse 
1. `sep = NA_character_` or `sep = "auto"` - don't know separator, try to guess about it.
1. `sep = any_single_character` - know separtor in advance, don't look for it.

Another option is to introduce new function, say, `fread_lines` which will return plain `character` and will have only a very few arguments.

**Comment by arunsrinivasan** [comment]
`sep="\n"` looks natural to me. Why does `sep=NULL` look more natural to you? Not sure what you mean by split/parse.. it still has to take care of embedded `\n`s, isn't it?

**Comment by arunsrinivasan** [comment]
Oh I see, `sep=NULL` reads the entire line as such (is your point). Hm, makes sense.

**Comment by arunsrinivasan** [comment]
I'm not yet sure about this. Will come back later.

**Comment by MichaelChirico** [comment]
Could you file an issue for this? I'd like to refer to an issue number rather than a PR number.

**Comment by dselivanov** [comment]
@MichaelChirico feel free to feel issue yourself if you really need it.

**Comment by MichaelChirico** [comment]
I think that's your responsibility, but fine.

From [Contribution Guidelines](https://github.com/Rdatatable/data.table/blob/master/Contributing.md):

> In your pull request's description, please state clearly... what FR or bug your PR addresses, along with the issue number.

**Comment by dselivanov** [comment]
Can I do anything else to improve this PR, so it could be merged?

**Comment by arunsrinivasan** [comment]
@dselivanov sorry for the delay. I've given it some thought. I still find `sep="\n"` to be more natural. Of course this implicitly assumes `"\r\n"` if necessary. Would it be possible to make this change?

To me, `sep=NULL` seems appropriate for the behaviour where the entire file is read in as a character vector of length=1, which is not a particularly useful feature, but just mentioning it to explain what `sep=NULL` makes me think of the behaviour.

**Comment by dselivanov** [comment]
@arunsrinivasan ok agree with your point about `sep=NULL`. I will add sep="\n" in documentation. Internally it will still work with `sep=NULL` (so auto detection of `\r\n` will work). Will this be appropriate?

**Comment by arunsrinivasan** [comment]
Sorry, I'm not following.

**Comment by arunsrinivasan** [comment]
Okay. As long as `sep=NULL` errors (current behaviour), it's fine.

**Comment by dselivanov** [comment]
See updated PR. Sorry for long update, totally forgot about it...

File: man/fread.Rd [hunk 1]
```diff
@@ -15,14 +15,14 @@ skip=0L, select=NULL, drop=NULL, colClasses=NULL,
 integer64=getOption("datatable.integer64"),         # default: "integer64"
 dec=if (sep!=".") "." else ",", col.names, 
 check.names=FALSE, encoding="unknown", quote="\"", 
-strip.white=TRUE, fill=FALSE, blank.lines.skip=FALSE, key=NULL, 
+strip.white = ifelse(sep == "\n", FALSE, TRUE), fill=FALSE, blank.lines.skip=FALSE, key=NULL,
```
Comment (jangorecki) [inline man/fread.Rd]: `ifelse` could be more useful for vectorized input, here `strip.white = sep!="\n"` or `strip.white = !identical(sep,"\n")` should be sufficient, isn't it? the latter one if we allow 0 length `sep` argument.

Comment (dselivanov) [inline man/fread.Rd]: good point!

**Comment by jangorecki** [comment]
does it wait for anything? assuming feedback was well addressed we could have this feature as it is pretty useful. @mattdowle @arunsrinivasan

**Comment by dselivanov** [comment]
I only kindly ask to merge it before 1.9.8 CRAN release :-).

21 июля 2016 г. 9:44 PM пользователь "Jan Gorecki" <notifications@github.com>
написал:

**Comment by arunsrinivasan** [comment]
@dselivanov if you could rebase and resolve the conflicts, I'll take a quick look and merge it in.

**Comment by jangorecki** [comment]
BTW. IMO `sep="\n"`is not a perfect option. If that feature is about to be used as `readLines` then there is just a single field in each line, so no separation really. `\n` refers to new line, which isn't another field, but another row. `sep` should refers to separation between fields in a row.

**Comment by arunsrinivasan** [comment]
@jangorecki EOF is always an implicit separator.

**Comment by jangorecki** [comment]
Makes more sense, so it is consistent to how `strsplit` would handle that. Looks good

**Comment by dselivanov** [comment]
I will try to resolve conflicts and update PR.

**Comment by dselivanov** [comment]
@arunsrinivasan rebased now.

File: inst/tests/tests.Rraw [hunk 2]
```diff
@@ -9108,6 +9108,12 @@ test_values <- c(53L, 53L, 52L, 1L, 52L, 1L, 1L,
 
 test(1702, isoweek(test_cases), test_values)
 
+# 
+lines <- fread("a,b\n ab,cd,ce\n abcdef\n hjkli \n", sep = "\n", header = T)[[1]]
+test(1591.1, lines, c(" ab,cd,ce", " abcdef", " hjkli ") )
```
Comment (jangorecki) [inline inst/tests/tests.Rraw]: test numbers can be adjusted to avoid duplicated id of tests

File: R/fread.R [hunk 3]
```diff
@@ -1,6 +1,15 @@
...
+fread <- function(input = "", sep = "auto", sep2 = "auto", nrows = -1L, header = "auto",
+                  na.strings = "NA", stringsAsFactors = FALSE, verbose = getOption("datatable.verbose"),
+                  autostart = 1L, skip = 0L, select = NULL, drop = NULL, colClasses = NULL,
+                  integer64 = getOption("datatable.integer64"), 
+                  dec=if (sep!=".") "." else ",", col.names,
+                  check.names = FALSE, encoding = "unknown", quote = "\"", 
+                  strip.white = !identical(sep,"\n"), 
+                  fill = FALSE, blank.lines.skip = FALSE, key = NULL, 
+                  showProgress = getOption("datatable.showProgress"),
+                  data.table = getOption("datatable.fread.datatable")) {
```
Comment (jangorecki) [inline R/fread.R]: it is more readable now :+1:

File: src/fread.c [hunk 4]
```diff
@@ -798,13 +802,16 @@ SEXP readfile(SEXP input, SEXP separg, SEXP nrowsarg, SEXP headerarg, SEXP nastr
     //   Auto detect separator, number of fields, and location of first row
     // ********************************************************************************************
     const char *seps;
-    if (isNull(separg)) {
+    if(!isNull(separg)) {
```
Comment (jangorecki) [inline src/fread.c]: ideally, this could have one extra space after `if`

File: inst/tests/tests.Rraw (hunk 2 above)
Comment (dselivanov) [inline inst/tests/tests.Rraw]: Nice catch. I have amended PR.

**Comment by dselivanov** [comment]
merge upstream to my PR request branch. 
@jangorecki I replaced your recent addition with`file` argument to `file=NULL`. It is not a good practice to have arguments without default value after arguments with default values.

**Comment by dselivanov** [comment]
@arunsrinivasan @jangorecki anything else I can do?

**Comment by arunsrinivasan** [comment]
Sorry, took a break. Seems fine but might be hard for this release as Matt seems to have almost wrapped up checking against dependent packages. Tagging @mattdowle just in case.

**Comment by dselivanov** [comment]
I don't think this can be easily merged - logic of `fread` moved forward a lot - closing.

**Pull request closed without merging**